app/iOSFS.m

```diff
@@ -0,0 +1,449 @@
+            }
+        });
+
+        pthread_mutex_lock(&open_mutex);
+        pthread_mutex_unlock(&open_mutex);
+
+        int posix_error = posixErrorFromNSError(error);
+        return posix_error ? ERR_PTR(posix_error) : openedFile;
+    } else {
+        return realfs_open_with_fdops(mount, path, flags, mode, &iosfs_fdops);
+    }
+}
+
+int iosfs_close(struct fd *fd) {
+    int err = realfs.close(fd);
+
+    if (fd->data) {
+        pthread_mutex_t *mutex = (pthread_mutex_t *)fd->data;
+        pthread_mutex_unlock(mutex);
+    }
+    return err;
+}
+
+static int iosfs_ask_for_url(NSURL **url) {
+    TerminalViewController *terminalViewController = currentTerminalViewController;
+
+    if (!terminalViewController)
+        return _ENODEV;
+
+    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+    DirectoryPickerDelegate *pickerDelegate = [[DirectoryPickerDelegate alloc] initWithMutex:&mutex];
+
+    dispatch_async(dispatch_get_main_queue(), ^(void) {
+        UIDocumentPickerViewController *picker = [[UIDocumentPickerViewController alloc] initWithDocumentTypes:@[ @"public.folder" ] inMode:UIDocumentPickerModeOpen];
+        picker.delegate = pickerDelegate;
+        [terminalViewController presentViewController:picker animated:true completion:nil];
+    });
+
+    pthread_mutex_lock(&mutex);
+    pthread_mutex_unlock(&mutex);
+
+    if (pickerDelegate.url == nil)
+        return _ENODEV;
+
+    *url = pickerDelegate.url;
+    return 0;
+}
+
+static int iosfs_mount(struct mount *mount) {
+    NSURL *url;
+
+    int err = iosfs_ask_for_url(&url);
+    if (err)
+        return err;
+
+    // Overwrite url & base path
+    mount->data = (void *)CFBridgingRetain(url);
+    mount->source = strdup([[url path] UTF8String]);
+
+    if ([url startAccessingSecurityScopedResource] == NO) {
+        CFBridgingRelease(mount->data);
+        return _EPERM;
+    }
+
+    return realfs.mount(mount);
+}
+
+static int iosfs_umount(struct mount *mount) {
+    NSURL *url = url_for_mount(mount);
+    [url stopAccessingSecurityScopedResource];
+    CFBridgingRelease(mount->data);
+
+    return 0;
+}
+
+static int iosfs_rename(struct mount *mount, const char *src, const char *dst) {
+    NSFileCoordinator *coordinator = [[NSFileCoordinator alloc] initWithFilePresenter:nil];
+    NSURL *src_url = url_for_path_in_mount(mount, src);
+    NSURL *dst_url = url_for_path_in_mount(mount, dst);
+
+    NSError *error;
+    __block int err;
+
+    [coordinator coordinateWritingItemAtURL:src_url options:NSFileCoordinatorWritingForMoving error:&error byAccessor:^(NSURL *url) {
+        [coordinator itemAtURL:url willMoveToURL:dst_url];
+        err = realfs.rename(mount, path_for_url_in_mount(mount, url, src), dst);
+        [coordinator itemAtURL:url didMoveToURL:dst_url];
+    }];
+
+    return combine_error(error, err);
+}
+
+static int iosfs_symlink(struct mount *mount, const char *target, const char *link) {
```

Comment:
There's a lot of repeated code in these functions. How much can be factored out into a new function that takes a block?